### PR TITLE
Fixes #1657 - Using 'dd' deletes the logical line and not the entire buffer.

### LINKED ISSUE #1657: [vi-mode] Using 'dd' should only delete the current logical line and not the entire buffer.
## Description of the new feature/enhancement

Using <kbd>d</kbd><kbd>d</kbd> currently deletes the entire buffer, which seems wrong when using multiline input.

### PR DESCRIPTION
**Pull request by springcomp**
# PR Summary

This PR includes an implementation of the <kbd>d</kbd><kbd>d</kbd> chord in VI-mode that deletes a (number of) logical line(s). This fixes the current behaviour which deletes the entire buffer, which seems wrong when using multiline input.

First, this PR includes a helper class that considers a "line" to consist of the first (optional) new-line character from the previous line up to the last non new-line character from the current line. This makes it easier to reason about chunks of _linewise_ text that have been deleted.

This PR also includes a refactor for the `EditItemDelete` class as well as an appropriate implementation for the new `EditItemDeleteLines` corresponding class.

-------

[update by @daxian-dbw] A new key binding <kbd>d</kbd><kbd>_</kbd> was added in the PR, the PR https://github.com/MicrosoftDocs/PowerShell-Docs/pull/6338 was opened in the doc repo to update the doc. 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/PowerShell/PSReadLine/pull/1658)

### CODE REVIEW

File: PSReadLine/KeyBindings.vi.cs [hunk 1]
```diff
@@ -231,6 +231,7 @@ private void SetDefaultViBindings()
                 { Keys.Space,           MakeKeyHandler( DeleteChar,                   "DeleteChar") },
                 { Keys._0,              MakeKeyHandler( BackwardDeleteLine,           "BackwardDeleteLine") },
                 { Keys.Uphat,           MakeKeyHandler( DeleteLineToFirstChar,        "DeleteLineToFirstChar") },
+                { Keys.Underbar,        MakeKeyHandler( DeleteLine,                   "DeleteLine") },
```
Comment (daxian-dbw) [inline PSReadLine/KeyBindings.vi.cs]: Can you please move this one right after the `Keys.D` (the other "DeleteLine")?

Also, this requires an update to the doc at https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/7.1/PSReadLine/About/about_PSReadLine.md#deleteline
/cc @sdwheeler

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: PSReadLine/StringBuilderLinewiseExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                if (buffer[position] == '\n')
+                {
+                    if (currentLine + 1 == lineIndex && !startPositionIdentified)
+                    {
+                        startPosition = position;
+                        startPositionIdentified = true;
+                    }
+
+                    currentLine++;
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: Good catch. Thank you for the feedback.

File: PSReadLine/StringBuilderLinewiseExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+                    if (currentLine == lineIndex + lineCount && !endPositionIdentified)
+                    {
+                        endPosition = position - 1;
+                        endPositionIdentified = true;
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: Done.

File: PSReadLine/StringBuilderLinewiseExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using System.Text;
+
+namespace Microsoft.PowerShell
+{
+    public class Range
+    {
+        public int Offset { get; set; }
+        public int Count { get; set; }
+    }
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs:15]: Done.

File: PSReadLine/StringBuilderLinewiseExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            }
+
+            return new Range
+            {
+                Offset = startPosition,
+                Count = endPosition - startPosition + 1,
+            };
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: Done.

File: PSReadLine/ReadLine.vi.cs [hunk 6]
```diff
@@ -722,11 +722,65 @@ public static void DeleteLineToFirstChar(ConsoleKeyInfo? key = null, object arg
...
+        /// return the offset to the deleted position.
+        /// </summary>
+        /// <returns></returns>
+        private static int DeleteLineImpl(int lineIndex, int lineCount)
+        {
+            var range = _singleton._buffer.GetRange(lineIndex, lineCount);
+
+            _clipboard.LinewiseRecord(_singleton._buffer.ToString(range.Offset, range.Count));
+
+            var deleteText = _singleton._buffer.ToString(range.Offset, range.Count);
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/ReadLine.vi.cs:767]: Done.

File: PSReadLine/ReadLine.vi.cs [hunk 7]
```diff
@@ -722,11 +722,65 @@ public static void DeleteLineToFirstChar(ConsoleKeyInfo? key = null, object arg
...
+            var deletePosition = range.Offset;
+            var anchor = _singleton._current;
+
+            _singleton._buffer.Remove(range.Offset, range.Count);
+
+            _singleton.SaveEditItem(EditItemDeleteLines.Create(
+                deleteText,
+                deletePosition,
+                anchor
+            ));
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/ReadLine.vi.cs]: Done.

File: PSReadLine/ReadLine.vi.cs [hunk 8]
```diff
@@ -722,11 +722,65 @@ public static void DeleteLineToFirstChar(ConsoleKeyInfo? key = null, object arg
...
+
+                // goto the first character of the first remaining logical line
+
+                var newCurrent = deletePosition + 1;
+
+                if (lineIndex + requestedLineCount >= lineCount)
+                {
+                    // if the delete operation has removed all the remaining lines
+                    // goto the first character of the previous logical line 
+
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/ReadLine.vi.cs]: OK.

File: PSReadLine/ReadLine.vi.cs [hunk 9]
```diff
@@ -722,11 +722,65 @@ public static void DeleteLineToFirstChar(ConsoleKeyInfo? key = null, object arg
...
+            {
+                var lineCount = _singleton.GetLogicalLineCount();
+                var lineIndex = _singleton.GetLogicalLineNumber() - 1;
+
+                TryGetArgAsInt(arg, out var requestedLineCount, 1);
+
+                var deletePosition = DeleteLineImpl(lineIndex, requestedLineCount);
+
+                // goto the first character of the first remaining logical line
+
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/ReadLine.vi.cs]: OK.

File: PSReadLine/UndoRedo.cs [hunk 10]
```diff
@@ -251,6 +259,29 @@ public override void Redo()
             }
         }
 
+        [DebuggerDisplay("DeleteLines '{_deletedString}' ({_deleteStartPosition}, Anchor: {_deleteAnchor})")]
+        class EditItemDeleteLines : EditItemDelete
+        {
+            private readonly int _deleteAnchor;
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/UndoRedo.cs:266]: Added:

```
// in linewise deletes, the _deleteAnchor represents the position
// of the cursor at the time delete was invoked. This is recorded
// so as to be restored when undoing the delete.
private readonly int _deleteAnchor;
```

File: PSReadLine/UndoRedo.cs [hunk 11]
```diff
@@ -224,18 +224,26 @@ public override void Undo()
...
             public static EditItem Create(string str, int position, Action<ConsoleKeyInfo?, object> instigator = null, object instigatorArg = null)
             {
                 return new EditItemDelete
-                {
-                    _deletedString = str,
-                    _deleteStartPosition = position,
-                    _instigator = instigator,
-                    _instigatorArg = instigatorArg
-                };
+                (
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/UndoRedo.cs]: Done.

File: test/StringBuilderLinewiseExtensionsTests.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+            var range = buffer.GetRange(1, 2);
+
+            // assert
+
+            Assert.Equal(5, range.Offset);
+            Assert.Equal(12, range.Count);
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (springcomp) [reply test/StringBuilderLinewiseExtensionsTests.cs]: Done.

File: PSReadLine/KeyBindings.vi.cs (hunk 1 above)
Comment (springcomp) [reply PSReadLine/KeyBindings.vi.cs]: [Added documentation](https://github.com/MicrosoftDocs/PowerShell-Docs/pull/6338) for the new <kbd>d</kbd><kbd>_</kbd> chord for the `DeleteLine` command.

File: PSReadLine/ViRegister.cs [hunk 13]
```diff
@@ -97,6 +97,11 @@ public int PasteAfter(StringBuilder buffer, int position)
                 {
                     var text = _text;
 
+                    if (text[0] != '\n')
```
[the remark this answers is not in the document]
Comment (springcomp) [reply PSReadLine/ViRegister.cs:100]: It does, indeed, relate to the <kbd>d</kbd><kbd>d</kbd> change.

Before my PR, there was no systemic way of thinking about what constitutes a _line_.
Lines are just chunks of text delimited by newline characters. But when handling linewise pastes, and deletes, one must adopt a convention as to whether the newline belongs to the previous line, the current line or to no line in particular.

The preliminary work for the <kbd>d</kbd><kbd>d</kbd> change is done in [the first commit](https://github.com/PowerShell/PSReadLine/pull/1658/commits/406a5090e37fc245401f2e7765192f2ee28723ff). There, I decided to arbitrarily consider a line as starting with a newline character and ending with a non-newline character.

I could have easily adopted any other convention but this one made the PR easier.

By doing that, this broke the [ViPasteAfterYankLine](https://github.com/PowerShell/PSReadLine/commit/f25a0892903c58cd4c490d5b1173d8cd926f1677#diff-9ada1b9035c9dc18120adf1223246c1fR210-R219) test, [initially introduced in a previous work](https://github.com/PowerShell/PSReadLine/commit/f25a0892903c58cd4c490d5b1173d8cd926f1677).

**Review by springcomp — COMMENTED** [review]
Addressed feedback.

File: PSReadLine/ViRegister.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply PSReadLine/ViRegister.cs:100]: Thanks for the clarification!

**Review by daxian-dbw — APPROVED** [review]
LGTM. Thanks @springcomp!

**Pull request merged**